GitHubReferences.ps1

```diff
@@ -0,0 +1,242 @@
+
+    .OUTPUTS
+        [PSCustomObject]
+        Details of the git reference in the given repository
+
+    .EXAMPLE
+        Get-GitHubReference -OwnerName Powershell -RepositoryName PowerShellForGitHub -Reference heads/master
+#>
+    [CmdletBinding(
+        SupportsShouldProcess,
+        DefaultParametersetName='Elements')]
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSShouldProcess", "",
+    Justification="Methods called within here make use of PSShouldProcess, and the switch is passed on to them inherently.")]
+    param(
+        [Parameter(ParameterSetName='Elements')]
+        [string] $OwnerName,
+
+        [Parameter(ParameterSetName='Elements')]
+        [string] $RepositoryName,
+
+        [Parameter(
+            Mandatory,
+            ParameterSetName='Uri')]
+        [string] $Uri,
+
+        [string] $Reference,
+
+        [string] $AccessToken,
+
+        [switch] $NoStatus
+    )
+
+    Write-InvocationLog -Invocation $MyInvocation
+
+    $elements = Resolve-RepositoryElements -BoundParameters $PSBoundParameters -DisableValidation
+    $OwnerName = $elements.ownerName
+    $RepositoryName = $elements.repositoryName
+
+    $telemetryProperties = @{
+        'OwnerName' = (Get-PiiSafeString -PlainText $OwnerName)
+        'RepositoryName' = (Get-PiiSafeString -PlainText $RepositoryName)
+        'ProvidedReference' = $PSBoundParameters.ContainsKey('Reference')
+    }
+
+    if ($OwnerName -xor $RepositoryName)
+    {
+        $message = 'You must specify both Owner Name and Repository Name.'
+        Write-Log -Message $message -Level Error
+        throw $message
+    }
+
+    if (-not [String]::IsNullOrEmpty($RepositoryName))
+    {
+        $uriFragment = "/repos/$OwnerName/$RepositoryName/git/refs"
+        $description =  "Getting all references for $RepositoryName"
+        if ($PSBoundParameters.ContainsKey('Reference'))
+        {
+            $uriFragment = $uriFragment + "/$Reference"
+            $description =  "Getting Reference $Reference for $RepositoryName"
+        }
+    }
+
+    $params = @{
+        'UriFragment' = $uriFragment
+        'Description' = $description
+        'AccessToken' = $AccessToken
+        'TelemetryEventName' = $MyInvocation.MyCommand.Name
+        'TelemetryProperties' = $telemetryProperties
+        'NoStatus' = (Resolve-ParameterWithDefaultConfigurationValue -Name NoStatus -ConfigValueName DefaultNoStatus)
+    }
+
+    try
+    {
+        return Invoke-GHRestMethodMultipleResult @params
+    }
+    catch
+    {
+        if ($PSBoundParameters.ContainsKey('Reference'))
+        {
+            Write-InteractiveHost "No reference named $Reference exists in repository $RepositoryName" -NoNewline -f Red
+            Write-Log -Level Error "No reference named $Reference exists in repository $RepositoryName"
+            return $null
+        }
+        Write-InteractiveHost "No references exist in repository $RepositoryName" -NoNewline -f Red

```

Comment:
In general, I think I want to avoid having individual functions using `Write-InteractiveHost`.  Writing it to `Error` via `Write-Log` should be enough to communicate to users, then `throw`.